File: Keep_Stock_Web/models/fornecedores.py
from core.crud_base import CrudBase
from core import database
from core.database import conectar_supabase
from core.validator import Validator
 
class Fornecedor(CrudBase):
    table = "fornecedores"
    primary_key = "id_fornecedor" 
    fields = [
        "nome",
        "cnpj",
        "telefone",
        "email",
        "cep",
        "imagem_url",
    ]

    def __init__(self, nome, cnpj, telefone, email, cep, imagem_url=None, id=None, id_fornecedor=None, **kwargs):
        self.nome = nome
        self.cnpj = cnpj
        self.telefone = telefone
        self.email = email
        self.cep = cep
        self.imagem_url = imagem_url
        self.id = id or id_fornecedor
        self.id_fornecedor = self.id
        self.db = conectar_supabase() # Pega o fornecedor do Supabase

    @classmethod
    def find_all(cls, order_by=None):
        # Sobrescreve find_all para usar id_fornecedor como chave primária
        if order_by is None:
            order_by = cls.primary_key
        supabase = conectar_supabase()
        response = supabase.table(cls.table).select("*").order(order_by).execute()
        return response.data

    @classmethod
    def find_by_id(cls, id):
        # Sobrescreve find_by_id para usar id_fornecedor como chave primária
        supabase = conectar_supabase()
        response = supabase.table(cls.table).select("*").eq(cls.primary_key, id).execute()
        return response.data[0] if response.data else None

    @classmethod
    def delete(cls, id):
        # Sobrescreve delete para usar id_fornecedor como chave primária
        supabase = conectar_supabase()
        response = supabase.table(cls.table).delete().eq(cls.primary_key, id).execute()
        return response.data

    def insert(self):
        # Insert
        supabase = conectar_supabase()
        dados = {campo: getattr(self, campo) for campo in self.fields}
        response = supabase.table(self.table).insert(dados).execute()
        return response.data

    def update(self, id):
        supabase = conectar_supabase()
        dados = {campo: getattr(self, campo) for campo in self.fields}
        response = supabase.table(self.table).update(dados).eq(self.primary_key, id).execute()
        return response.data

    def validate(self):
        erros = [
            Validator.required(self.nome, "nome"),
            Validator.required(self.cnpj, "cnpj"),
            Validator.required(self.telefone, "telefone"),
            Validator.required(self.email, "email"),
            Validator.required(self.cep, "cep"),
        ]
        return [erro for erro in erros if erro]
                    
    @classmethod
    def has_related_records(cls, id):
        # Verifica se o fornecedor possui registros vinculados (ex.: lotes)
        supabase = conectar_supabase()
        try:
            response = supabase.table("lotes").select("*", count="exact").eq("id_fornecedor", id).execute()
            if response.count and response.count > 0:
                return True
            return False
        except Exception as e:
            print(f"Erro ao verificar registros relacionados: {e}")
            return False

    @classmethod
    def safe_delete(cls, id):
        supabase = conectar_supabase()
        # Busca produto e verifica se ele existe
        fornecedor = cls.find_by_id(id)
        if not fornecedor:
            raise ValueError("fornecedor não encontrado.")
        if cls.has_related_records(id):
            raise ValueError("Não é possível excluir o fornecedor porque possui tabelas vinculadas.")
        cls.delete(id)

    @classmethod
    def buscar_por_fornecedor(cls, id_fornecedor):
        # Busca um fornecedor no banco e retorna uma instância da classe
        supabase = conectar_supabase()
        response = supabase.table("fornecedores").select("*").eq(cls.primary_key, id_fornecedor).execute()
        
        if response.data:
            d = response.data[0]
            return cls(
                id=d.get('id_fornecedor'),
                nome=d.get('nome'),
                cnpj=d.get('cnpj'),
                telefone=d.get('telefone'),
                email=d.get('email'),
                cep=d.get('cep'),
                imagem_url=d.get('imagem_url'),
            )
        return None
    
    def inserir_fornecedor(self):
        # Insere o fornecedor usando o método da classe base
        supabase = conectar_supabase()
        dados = {campo: getattr(self, campo) for campo in self.fields}
        response = supabase.table(self.table).insert(dados).execute()
        return response.data

File: Mobile_Keep_Stock/src/services/fornecedores.js
import { criarCrud } from './crudBase';
import { filtrarPorTermo } from './helpers';
import { BUCKET_FORNECEDORES, deletarImagem, uploadImagem } from './imagemStorage';
import { Validator } from './validator';

// ROTAS DE FORNECEDORES
// Espelha /fornecedores e /fornecedor/salvar, /fornecedor/atualizar e /fornecedor/excluir

export const fornecedoresCrud = criarCrud('fornecedores', 'id_fornecedor');

// Normaliza os campos vindos do formulário (espelha get_fornecedor_form)
export function montarFornecedor(dados) {
  return {
    nome: (dados.nome || '').trim(),
    cnpj: (dados.cnpj || '').trim(),
    telefone: (dados.telefone || '').trim(),
    email: (dados.email || '').trim(),
    cep: (dados.cep || '').trim(),
  };
}

// Espelha Fornecedor.validate()
export function validarFornecedor(fornecedor) {
  const erros = [
    Validator.required(fornecedor.nome, 'nome'),
    Validator.required(fornecedor.cnpj, 'cnpj'),
    Validator.required(fornecedor.telefone, 'telefone'),
    Validator.required(fornecedor.email, 'email'),
    Validator.required(fornecedor.cep, 'cep'),
  ];

  return erros.filter(Boolean);
}

export async function listarFornecedores(termo = '') {
  const fornecedores = await fornecedoresCrud.findAll('nome');

  return filtrarPorTermo(fornecedores, termo, [
    'nome',
    'cnpj',
    'telefone',
    'email',
    'cep',
  ]);
}

export async function buscarFornecedor(id) {
  return fornecedoresCrud.findById(id);
}

export async function salvarFornecedor(dados, imagem = null) {
  const fornecedor = montarFornecedor(dados);
  let imagemEnviada = null;

  try {
    imagemEnviada = imagem ? await uploadImagem(imagem, BUCKET_FORNECEDORES) : null;
    fornecedor.imagem_url = imagemEnviada;

    const erros = validarFornecedor(fornecedor);

    if (erros.length) throw new Error(erros.join(' '));

    return await fornecedoresCrud.insert(fornecedor);
  } catch (erro) {
    if (imagemEnviada) await deletarImagem(imagemEnviada, BUCKET_FORNECEDORES);

    throw erro;
  }
}

export async function atualizarFornecedor(id, dados, imagem = null) {
  const fornecedorAtual = await fornecedoresCrud.findById(id);

  if (!fornecedorAtual) throw new Error('Fornecedor não encontrado.');

  const fornecedor = montarFornecedor(dados);
  fornecedor.imagem_url = fornecedorAtual.imagem_url;

  let novaImagem = null;

  try {
    if (imagem) {
      novaImagem = await uploadImagem(imagem, BUCKET_FORNECEDORES);
      fornecedor.imagem_url = novaImagem;
    }

    const erros = validarFornecedor(fornecedor);

    if (erros.length) throw new Error(erros.join(' '));

    const atualizado = await fornecedoresCrud.update(id, fornecedor);

    if (novaImagem && fornecedorAtual.imagem_url) {
      await deletarImagem(fornecedorAtual.imagem_url, BUCKET_FORNECEDORES);
    }

    return atualizado;
  } catch (erro) {
    if (novaImagem) await deletarImagem(novaImagem, BUCKET_FORNECEDORES);

    throw erro;
  }
}

// Verifica se o fornecedor possui registros vinculados (ex.: lotes)
export async function fornecedorPossuiVinculos(id) {
  try {
    const lotes = await criarCrud('lotes').contarPor('id_fornecedor', id);

    return lotes > 0;
  } catch (erro) {
    console.log(`Erro ao verificar registros relacionados: ${erro}`);
    return false;
  }
}

// Espelha Fornecedor.safe_delete()
export async function excluirFornecedor(id) {
  const fornecedor = await fornecedoresCrud.findById(id);

  if (!fornecedor) throw new Error('Fornecedor não encontrado.');

  if (await fornecedorPossuiVinculos(id)) {
    throw new Error(
      'Não é possível excluir o fornecedor porque possui tabelas vinculadas.'
    );
  }

  await fornecedoresCrud.delete(id);

  if (fornecedor.imagem_url) {
    await deletarImagem(fornecedor.imagem_url, BUCKET_FORNECEDORES);
  }
}
